crates/parameters/src/lib.rs

```diff
@@ -459,3 +442,208 @@ where
     };
     init_storage(&params, storage)
 }
+
+/// Return an estimate of the maximum time taken to decide a block,
+/// by sourcing block headers from up to `num_blocks_to_read`.
+pub fn estimate_max_block_time_from_blocks<S>(
+    storage: &S,
+    last_block_height: BlockHeight,
+    num_blocks_to_read: u64,
+) -> namada_storage::Result<Option<DurationSecs>>
+where
+    S: StorageRead,
+{
+    let ending_height = last_block_height.0;
+    let beginning_height =
+        ending_height.saturating_sub(num_blocks_to_read.saturating_sub(1));
+
+    let block_timestamps = {
+        let mut ts = Vec::with_capacity(
+            usize::try_from(num_blocks_to_read).into_storage_result()?,
+        );
+
+        for height in beginning_height..=ending_height {
+            let Some(block_header) =
+                storage.get_block_header(BlockHeight(height))?
+            else {
+                break;
+            };
+            ts.push(block_header.time);
+        }
+
+        ts
+    };
+
+    Ok(block_timestamps
+        .windows(2)
+        // NB: compute block time
+        .map(|ts| {
+            #[allow(clippy::arithmetic_side_effects)]
+            {
+                ts[1] - ts[0]
+            }
+        })
+        .max())
+}
+
+/// Return an estimate of the maximum time taken to decide a block,
+/// based on chain parameters.
+pub fn estimate_max_block_time_from_parameters<S>(
+    storage: &S,
+) -> namada_storage::Result<DurationSecs>
+where
+    S: StorageRead,
+{
+    let EpochDuration {
+        min_num_of_blocks,
+        min_duration: DurationSecs(min_duration),
+    } = read_epoch_duration_parameter(storage)?;
+
+    checked!(min_duration / min_num_of_blocks)
+        .map(|block_time| {
+            // NB: fallback to one block per epoch
```

Comment:
On block per second here?

Reply:
one block per `$minimum-epoch-duration`, which you compute from `$epochs-per-year`